### ISSUE TITLE
Unhelpful error message when using non-string as key

### ISSUE THREAD

**Issue by tomhicks-bsf** [NodeRedis/node_redis#1013]
* **Version**: 2.5.2
* **Platform**: any
* **Description**:

A change was made around 2.5.0 which is described in the changelog as:
> Fixed do not run toString on an array argument and throw a "invalid data" error instead

I can't find the commit that relates to this, as it's quite hard to find source code that is no longer in the repo.

This is described as a non-breaking change (in the changelog) but it is a very breaking change. I can see why that change was made, but the error that goes with it is extremely unhelpful:
```
TypeError: invalid data
    at Socket.write (net.js:635:11)
    at RedisClient.write (/opt/nodeservice/node_modules/redis/index.js:901:43)
    at RedisClient.send_command (/opt/nodeservice/node_modules/redis/index.js:870:22)
    at RedisClient.(anonymous function).RedisClient.(anonymous function) (/opt/nodeservice/node_modules/redis/lib/commands.js:45:25)
```

What would be much better is an error that said "Key must be a string" or something to that effect. I'm happy to make this change if someone could at least point me to where I need to make this check. I think it might be in lib/multi.js:39 ish.

[gitter]: https://gitter.im/NodeRedis/node_redis?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge

**Comment by BridgeAR** [comment]
@tomhicks-bsf I totally understand that this is not a very good error message especially, as that code earlier "worked" (that was more by chance though and not guaranteed ;) ).

You can still use non-string values but only numbers, dates and buffers (and null and undefined, but those are going to throw from v.3 on too and are deprecated right now). Handling this with a better error is likely going to decrease the performance for saving buffers a tiny bit though :/. But I agree that this should be changed and I actually think about switching the error to a deprecation, even though most people who used arrays as arguments have already run into that error.

The corresponding [line](https://github.com/NodeRedis/node_redis/blob/master/index.js#L808) is in the send_command function and there has to be a check for `Buffer.isBuffer()` plus a new else case handling all the data types that are not supported (and therefor should either receive a good error message or a deprecation message).

**Comment by jbt** [comment]
Just adding my two cents here that the fact that this now throws synchronously (rather than calling back with an error or triggering an error event) feels Very Breaking to me - I just got caught out where using nested objects for hmset now crashes the process rather than "gracefully" storing `"[object Object]"` instead.

In my case, it's easy enough to put some validation in a few places, but it does feel awkward having to do that for a minor version update labelled as "non-breaking"

**Comment by tomhicks-bsf** [comment]
Good stuff :+1:

**Comment by BridgeAR** [comment]
Fixed and released as 2.5.3

**Comment by jbt** [comment]
Nice one! :+1:
